src/btree/bt_cursor.c

```diff
@@ -580,7 +588,15 @@ __wt_btcur_search(WT_CURSOR_BTREE *cbt)
                 if (F_ISSET(cursor, WT_CURSTD_KEY_ONLY))
                     valid = true;
                 else
-                    WT_ERR(__wt_cursor_valid(cbt, cbt->tmp, WT_RECNO_OOB, &valid));
+                    /*
+                     * Whenever the cursor doesn't found an exact match or found an exact match in
+                     * the insert list, it is possible that temporary key of the cursor and the
+                     * cursor position slot will differ. Pass the key as NULL instead of the cursor
+                     * temporary key to let the key to be generated internally from cursor position
+                     * slot to access the history store.
+                     */
+                    WT_ERR(__wt_cursor_valid(
+                      cbt, (cbt->ins != NULL ? NULL : cbt->tmp), WT_RECNO_OOB, &valid));
```

Comment:
We've had problems with `cbt.tmp` before, where callers assumed it was set when it wasn't. What would you think of making it so `cbt.tmp` is always set if `cbt.compare == 0`? It should make it easier to reason about this stuff in the future.

I think it's a reasonably easy change, and if I'm correct, it's a simpler change overall.
```
diff --git a/src/btree/bt_ret.c b/src/btree/bt_ret.c
index b6a35e6c7..ab0fcc4f5 100644
--- a/src/btree/bt_ret.c
+++ b/src/btree/bt_ret.c
@@ -24,20 +24,12 @@ __key_return(WT_CURSOR_BTREE *cbt)
     cursor = &cbt->iface;
 
     if (page->type == WT_PAGE_ROW_LEAF) {
-        /* If the cursor references a WT_INSERT item, take its key. */
-        if (cbt->ins != NULL) {
-            cursor->key.data = WT_INSERT_KEY(cbt->ins);
-            cursor->key.size = WT_INSERT_KEY_SIZE(cbt->ins);
-            return (0);
-        }
-
         /*
-         * If not in an insert list and there's an exact match, the row-store search function built
-         * the key we want to return in the cursor's temporary buffer. Swap the cursor's search-key
-         * and temporary buffers so we can return it (it's unsafe to return the temporary buffer
-         * itself because our caller might do another search in this table using the key we return,
-         * and we'd corrupt the search key during any subsequent search that used the temporary
-         * buffer).
+         * If there's an exact match, the row-store search function built the key we want to return
+         * in the cursor's temporary buffer. Swap the cursor's search-key and temporary buffers so
+         * we can return it (it's unsafe to return the temporary buffer itself because our caller
+         * might do another search in this table using the key we return, and we'd corrupt the
+         * search key during any subsequent search that used the temporary buffer).
          */
         if (cbt->compare == 0) {
             tmp = cbt->row_key;
diff --git a/src/btree/row_srch.c b/src/btree/row_srch.c
index 310659bb1..ca8ce844c 100644
--- a/src/btree/row_srch.c
+++ b/src/btree/row_srch.c
@@ -586,7 +586,8 @@ leaf_match:
 
     /*
      * Test for an append first when inserting onto an insert list, try to catch cursors repeatedly
-     * inserting at a single point.
+     * inserting at a single point, then search the insert list. If we find an exact match, copy the
+     * key into the temporary buffer, our callers expect to find it there.
      */
     if (insert) {
         WT_ERR(__search_insert_append(session, cbt, ins_head, srch_key, &done));
@@ -594,7 +595,10 @@ leaf_match:
             return (0);
     }
     WT_ERR(__wt_search_insert(session, cbt, ins_head, srch_key));
-
+    if (cbt->compare == 0) {
+       cursor->tmp.data = WT_INSERT_KEY(cbt->ins);
+        cursor->tmp.size = WT_INSERT_KEY_SIZE(cbt->ins);
+    }
     return (0);
```

Reply:
Thanks, I fixed the code and also updated the code paths where the cbt->tmp is used.